src/order/bounded_lattice.lean

```diff
@@ -1100,3 +1103,92 @@ is_compl.of_eq bot_inf_eq sup_top_eq
 
 lemma is_compl_top_bot [bounded_lattice α] : is_compl (⊤ : α) ⊥ :=
 is_compl.of_eq inf_bot_eq top_sup_eq
+
+end is_compl
+
+section atoms
+
+section is_atom
+
+variable [order_bot α]
+
+/-- An atom of an `order_bot` is an element with no other element between it and `⊥`,
+  which is not `⊥`. -/
+def is_atom (a : α) : Prop := a ≠ ⊥ ∧ (∀ b, b < a → b = ⊥)
+
+lemma eq_bot_or_eq_of_le_atom {a b : α} (ha : is_atom a) (hab : b ≤ a) : b = ⊥ ∨ b = a :=
+or.imp_left (ha.2 b) (lt_or_eq_of_le hab)
+
+end is_atom
+
+section is_coatom
+
+variable [order_top α]
+
+/-- A coatom of an `order_top` is an element with no other element between it and `⊤`,
+  which is not `⊤`. -/
+def is_coatom (a : α) : Prop := a ≠ ⊤ ∧ (∀ b, a < b → b = ⊤)
+
+lemma eq_top_or_eq_of_coatom_le {a b : α} (ha : is_coatom a) (hab : a ≤ b) : b = ⊤ ∨ b = a :=
+or.imp (ha.2 b) eq_comm.2 (lt_or_eq_of_le hab)
+
+end is_coatom
+
+variables [bounded_lattice α] {a : α}
+
+lemma is_atom_iff_is_coatom_dual : is_atom a ↔ @is_coatom (order_dual α) _ a := iff.refl _
```

Comment:
I think this might be better without the `@`s
```suggestion
lemma is_atom_iff_is_coatom_dual : is_atom a ↔ is_coatom (order_dual.to_dual a) := iff.refl _
```

Reply:
Can't do that without moving the file, as `order_dual.to_dual` is defined in a file that imports this one.
Is it worth it?

Reply:
Ah, there's an open issue about that. Does this work?
```suggestion
lemma is_atom_iff_is_coatom_dual : is_atom a ↔ is_coatom (a : order_dual α) := iff.refl _
```

Reply:
No, this casts `a` back to `α` for whatever reason. For instance,
`lemma foo : is_atom a ↔ is_atom (a : order_dual α) := iff.refl _` works.

Reply:
Hmm, the issue I'm thinking of is closed (#4758). Perhaps we need a new one.

Reply:
>  as order_dual.to_dual is defined in a file that imports this one.

are you sure? [the docs](https://leanprover-community.github.io/mathlib_docs/order/order_dual.html) seem to suggest that it imports only `order.basic`, and this is `order.bounded_lattice`.

Reply:
It imports `data.equiv.basic` which imports `data.set.function` which impports `data.set.basic` which imports `order.boolean_algebra`, and we get back to `order.bounded_lattice`.

Reply:
I could always move the contents of this PR to a new file, but I don't think it's worth it unless you or Johan really thinks it is.

Reply:
Can these lemmas go in `order_dual.lean` (with a TODO comment perhaps) until we sort out that mess?

Reply:
I do however expect it's worth it to have all the basics of dealing with `order_dual` available before any of the lattice ideas, but really what I'd want is a dualizing tactic.

Reply:
Perhaps we should open an issue to remove the uses of `data.set.function` from `data.equiv.basic`, moving them to a new `data.equiv.set`.

Reply:
Sounds good. I'm also just moving everything to a new file, I changed my mind.